File: leetcode/17.letter-combinations-of-a-phone-number.py
# @leet start
class Solution:
    def letterCombinations(self, digits: str) -> list[str]:
        LETTERS = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        res: list[str] = []
        curr: list[str] = []

        def backtrack(i: int) -> None:
            if i == len(digits):
                res.append("".join(curr))

                return

            for ch in LETTERS[digits[i]]:
                curr.append(ch)
                backtrack(i + 1)
                curr.pop()

            return

        backtrack(0)

        return res


# @leet end

File: leetcode/17.letter-combinations-of-a-phone-number.ts
// @leet start
function letterCombinations(digits: string): string[] {
  const digitToChars = {
    "2": ["a", "b", "c"],
    "3": ["d", "e", "f"],
    "4": ["g", "h", "i"],
    "5": ["j", "k", "l"],
    "6": ["m", "n", "o"],
    "7": ["p", "q", "r", "s"],
    "8": ["t", "u", "v"],
    "9": ["w", "x", "y", "z"],
  };

  const result: string[] = [];
  const curr: string[] = [];

  const backtrack = (i: number) => {
    if (i === digits.length) {
      result.push(curr.join(""));
      return;
    }

    for (const char of digitToChars[digits[i]]) {
      curr.push(char);
      backtrack(i + 1);
      curr.pop();
    }
  };

  backtrack(0);

  return result;
}
// @leet end
